### DISCUSSION TITLE
libsndfile for android with external libs support available

### DISCUSSION THREAD

**Discussion by scar20** [libsndfile/libsndfile#983] [General]
#### Complete build of libsndfile with Xiph libs and mpg support across 4 abi including the testsuite for android using the NDK.
https://github.com/scar20/libsndfile/tree/android_testing
for cross compiling on linux
ABI: - **`armeabi-v7a`** - **`arm64-v8a`** - **`x86_64`** - **`x86`**

Posting here for review before a formal pull request as the changes involve the testsuite. Comments welcome.

Changes:
- Addition:  [`cmake/CMakeAndroidTesting.cmake`](https://github.com/scar20/libsndfile/blob/android_testing/cmake/CMakeAndroidTesting.cmake).
- A switch in `CMakeLists.txt` to enable Android tests pointing to the above file.
- Check for condition `defined (__ANDROID__)` to skip one impossible test in `tests/misc_test.c`.
- New directory [`Script/AndroidScripts/`](https://github.com/scar20/libsndfile/tree/android_testing/Scripts/AndroidScripts) containing all the shell scripts for libsndfile and external libs build.
- An [`android_build_with_extlibs.md`](https://github.com/scar20/libsndfile/blob/android_testing/Scripts/AndroidScripts/Android_build_with_extlibs.md) in the above directory.
- A link in README.md to the above file.

Binaries of libsndfile packaged for Android Studio is available in the [`Release`](https://github.com/scar20/libsndfile/releases) section.
Precompiled binaries of the external libraries are also available in this section.

The testsuite have been tested under arm64-v8a, untested on the others architectures (will test those with emulators).
Attached is tests results for arm64-v8a:
[tests_result-sndfile-arm64-v8a.txt](https://github.com/libsndfile/libsndfile/files/13470425/tests_result-sndfile-arm64-v8a.txt)

**Comment by scar20** [comment]
Finally being able to test all Abi's:
[tests_result-sndfile-arm64-v8a.txt](https://github.com/libsndfile/libsndfile/files/13518962/tests_result-sndfile-arm64-v8a.txt)
[tests_result-sndfile-armeabi-v7a.txt](https://github.com/libsndfile/libsndfile/files/13518963/tests_result-sndfile-armeabi-v7a.txt)
[tests_result-sndfile-x86.txt](https://github.com/libsndfile/libsndfile/files/13518964/tests_result-sndfile-x86.txt)
[tests_result-sndfile-x86_64.txt](https://github.com/libsndfile/libsndfile/files/13518965/tests_result-sndfile-x86_64.txt)

**Reply by mattavent** [reply]
Hello! Does this support ogg/vorbis linkage? CUrrently having some issues getting vorbis functionality on android.

**Reply by scar20** [reply]
Yes, in theory,
I did not use ogg/vorbis myself but it had pass all the test. So you should be able to open a vorbis file in the same transparent way as any other libsndfile format.
You need to be in the 'android_testing' branch from my clone repo. If you want to compile yourself you'll find all the instructions in the 'Scripts/AndroidScripts/Android_build_with_extlibs.md. There is also pre-compiled libs available in the 'release' section. I'm late on the latest release of libsndfile though, I should update and sync...

**Reply by mattavent** [reply]
Great - I'm currently having a play with this to see if I can get it working. Are your precompiled releases of libsndfile compiled with ogg/vorbis support (presumabe dynamic linkage). The issue I've run into elsewhere is that cross-compiling sndfile with ogg/vorbis support seems to always fail no matter what information I provide it and no matter whether I use static or dynamic linkage. It's a bit mysterious...

**Reply by mattavent** [reply]
Ok I can confirm that running your scripts reports that it buildz a .so correctly but at runtime the ogg/vorbis funtionality is not available and outputs the following error "libsndfile was not built with OGG/Vorbis support or invalid format!". This is the same result I get with my own libsndfile build script. Any thoughts?

**Comment by mattavent** [comment]
your precompiled libsndfile.so does appear to be working with ogg/vorbis support tho - it's also 8MB compared to mine being 2.2MB -would be interesting to understand where the differences lie. Do you have a build dockerfile that encapsulates the environment for your .so builds please? with that I stand a better chance of unpicking what's going wrong :)

**Reply by mattavent** [reply]
(Oh the 8MB vs 2.2MB size difference will likely be me only statically linking ogg+vorbis (or none if its broken) vs you linking all the codec libs - FLAC etc)

**Reply by scar20** [reply]
You cannot build for ogg/vorbis only, you need the 4 ogg vorbis flac opus libs ( the 'xiph' switch in libsndfile build script). Those 4 are required by the internal libsndfile build script. If any of these are missing, it switch off those libs support. This probably why your build didnt work.
The provided externals libs are static libs so will be part of the resulting libsndfile.so sheared build.
You can set the build to target only the abi you'll use (most probably `arm64-v8a`). You can set this from the line:
`abi_list=("armeabi-v7a" "arm64-v8a" "x86" "x86_64")`
at the top of the main script `libsndfile/Scripts/AndroidScripts/build_libsndfile.sh). Just remove those you dont need.
You can also omit mpg support if you dont need it (mpg123, mp3lame)

Try to build from source following the instructions in the 'Scripts/AndroidScripts/Android_build_with_extlibs.md'
Use the external pre-compiled libs provided. Make sure you put them at the top level dir:
Top/Android_build_with_extlibs.md
Top/libsndfile/ source
Top/ogg/ pre-compiled
Top/flac/  pre-compiled
Top/vorbis/  pre-compiled
Top/opus/  pre-compiled

It should go smoothly (crossing my fingers). Let me know.

Do you have a build dockerfile - no I dont have that

**Reply by mattavent** [reply]
Aha interesting regarding needing all the libs! I hadn't spotted that. That strikes me as an odd choice on their part - no reason to bloat the library for things you don't need. Anyway - super helpful thanks :) I'll run another test with all of them asap.
